WhateverGreen/kern_rad.cpp

```diff
@@ -183,7 +192,108 @@ void RAD::processKernel(KernelPatcher &patcher, DeviceInfo *info) {
+	callbackRAD->dcLinkPtr = that; // save the dc link ptr for later backlight control
+	callbackRAD->getMaxBrightnessFromDisplay(); // read max brightness value from IOReg
+	return FunctionCast(wrapDce110EdpBacklightControl, callbackRAD->orgDce110EdpBacklightControl)(that, on_or_off, a3, a4, a5, a6);
+}
+
+IOReturn RAD::wrapAMDRadeonX6000AmdRadeonFramebufferSetAttribute(IOService *framebuffer, IOIndex connectIndex, IOSelect attribute, uintptr_t value) {
+	IOReturn ret = FunctionCast(wrapAMDRadeonX6000AmdRadeonFramebufferSetAttribute, callbackRAD->orgAMDRadeonX6000AmdRadeonFramebufferSetAttribute)(framebuffer, connectIndex, attribute, value);
+	if (attribute == (UInt32)'bklt') {
+		if (callbackRAD->maxPwmBacklightLvl != 0) {
+			// set the backlight of AMD navi10 driver
+			callbackRAD->curPwmBacklightLvl = value;
+			double btlper = (double)callbackRAD->curPwmBacklightLvl / callbackRAD->maxPwmBacklightLvl;
+			if (btlper < 0) {
+				btlper = 0;
+			}
+			if (btlper > 1.0) {
+				btlper = 1.0;
+			}
+			int pwmval = (int)(btlper * 0xFF) << 8;
+			if (pwmval >= 0xFF00) {
+				pwmval = 0x1FF00;
+			}
+
+			if (callbackRAD->dcLinkPtr) {
+				callbackRAD->wrapDcLinkSetBacklightLevel(callbackRAD->dcLinkPtr, pwmval, 0);
+			} else {
+				SYSLOG("igfx", "wrapAMDRadeonX6000AmdRadeonFramebufferSetAttribute null dc link");
+			}
+		} else {
+			SYSLOG("igfx", "wrapAMDRadeonX6000AmdRadeonFramebufferSetAttribute zero maxPwmBacklightLvl");
+		}
+		
+		ret = 0;
+	}
+	return ret;
+}
+
+IOReturn RAD::wrapAMDRadeonX6000AmdRadeonFramebufferGetAttribute(IOService *framebuffer, IOIndex connectIndex, IOSelect attribute, uintptr_t * value) {
+	IOReturn ret = FunctionCast(wrapAMDRadeonX6000AmdRadeonFramebufferGetAttribute, callbackRAD->orgAMDRadeonX6000AmdRadeonFramebufferGetAttribute)(framebuffer, connectIndex, attribute, value);
+	if (attribute == (UInt32)'bklt') {
+		// enable the backlight feature of AMD navi10 driver
+		*value = callbackRAD->curPwmBacklightLvl;
+		ret = 0;
+	}
+	return ret;
+}
+
 bool RAD::processKext(KernelPatcher &patcher, size_t index, mach_vm_address_t address, size_t size) {
+	if (kextRadeonX6000Framebuffer.loadIndex == index) {
+		KernelPatcher::RouteRequest requests[] = {
+			{"_dce110_edp_backlight_control", wrapDce110EdpBacklightControl, orgDce110EdpBacklightControl},
+			{"__ZN35AMDRadeonX6000_AmdRadeonFramebuffer25setAttributeForConnectionEijm", wrapAMDRadeonX6000AmdRadeonFramebufferSetAttribute, orgAMDRadeonX6000AmdRadeonFramebufferSetAttribute},
+			{"__ZN35AMDRadeonX6000_AmdRadeonFramebuffer25getAttributeForConnectionEijPm", wrapAMDRadeonX6000AmdRadeonFramebufferGetAttribute, orgAMDRadeonX6000AmdRadeonFramebufferGetAttribute},
+			{"_dc_link_set_backlight_level", wrapDcLinkSetBacklightLevel, orgDcLinkSetBacklightLevel},
+		};
+
+		if (!patcher.routeMultiple(index, requests, address, size, true, true))
+			SYSLOG("igfx", "DBG: Failed to route amd gpu tracing.");
```

Comment:
drop DBG, this does not look like debug to me. We also have DBGLOG macro for debug logs.